File: RocketPocket/ios/Sources/DirectionPad.swift
import SwiftUI

/// One cell of the 3×3 pad. A single upward arrow rotated into place keeps all eight
/// directions visually identical.
private struct Direction {
    let command: Character
    let label: String
    let rotation: Double
    let accent: Color
}

private let directions: [Character: Direction] = [
    Command.forwardLeft:   Direction(command: Command.forwardLeft,   label: "FWD LEFT",   rotation: -45,  accent: Palette.neonOrange),
    Command.forward:       Direction(command: Command.forward,       label: "FORWARD",    rotation: 0,    accent: Palette.neonCyan),
    Command.forwardRight:  Direction(command: Command.forwardRight,  label: "FWD RIGHT",  rotation: 45,   accent: Palette.neonOrange),
    Command.left:          Direction(command: Command.left,          label: "LEFT",       rotation: -90,  accent: Palette.neonCyan),
    Command.right:         Direction(command: Command.right,         label: "RIGHT",      rotation: 90,   accent: Palette.neonCyan),
    Command.backwardLeft:  Direction(command: Command.backwardLeft,  label: "BACK LEFT",  rotation: -135, accent: Palette.neonOrange),
    Command.backward:      Direction(command: Command.backward,      label: "BACKWARD",   rotation: 180,  accent: Palette.neonCyan),
    Command.backwardRight: Direction(command: Command.backwardRight, label: "BACK RIGHT", rotation: 135,  accent: Palette.neonOrange),
]

/// The eight-way drive pad with the emergency stop in the centre cell.
struct DirectionPad: View {
    let enabled: Bool
    let onPress: (Character) -> Void
    let onRelease: () -> Void
    let onEmergencyStop: () -> Void

    private let rows: [[Character?]] = [
        [Command.forwardLeft, Command.forward, Command.forwardRight],
        [Command.left, nil, Command.right],
        [Command.backwardLeft, Command.backward, Command.backwardRight],
    ]

    var body: some View {
        VStack(spacing: 10) {
            ForEach(0..<3, id: \.self) { row in
                HStack(spacing: 10) {
                    ForEach(0..<3, id: \.self) { column in
                        if let command = rows[row][column], let direction = directions[command] {
                            ControlButton(
                                label: direction.label,
                                systemImage: "arrow.up",
                                rotation: direction.rotation,
                                accent: direction.accent,
                                enabled: enabled,
                                onPress: { onPress(direction.command) },
                                onRelease: onRelease
                            )
                        } else {
                            EmergencyStopButton(onStop: onEmergencyStop)
                        }
                    }
                }
            }
        }
    }
}

/// A hold-to-drive button.
///
/// `DragGesture(minimumDistance: 0)` is how UIKit's touch-down/up/cancel maps onto SwiftUI:
/// `onChanged` fires the moment a finger lands, and `onEnded` fires when it lifts — while
/// `.onDisappear` and the gesture being interrupted both funnel into the same release handler.
/// There is no path through this gesture that leaves the car driving.
///
/// The gesture stays active even when `enabled` is false: the view model answers a press with
/// "Connect to Rocket Pocket first" instead of silently ignoring it.
struct ControlButton: View {
    let label: String
    let systemImage: String
    var rotation: Double = 0
    let accent: Color
    let enabled: Bool
    let onPress: () -> Void
    let onRelease: () -> Void

    @State private var pressed = false

    private var container: Color {
        if pressed && enabled { return accent.opacity(0.30) }
        return enabled ? Palette.carbonSurfaceHigh : Palette.carbonSurface
    }

    private var content: Color {
        if pressed && enabled { return accent }
        return enabled ? Palette.textPrimary : Palette.textDisabled
    }

    var body: some View {
        VStack(spacing: 3) {
            Image(systemName: systemImage)
                .font(.system(size: 30, weight: .bold))
                .rotationEffect(.degrees(rotation))
            Text(label)
                .font(.system(size: 9, weight: .medium))
                .tracking(1)
                .lineLimit(1)
                .minimumScaleFactor(0.7)
        }
        .foregroundColor(content)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(
            RoundedRectangle(cornerRadius: 18)
                .fill(container)
                .shadow(color: pressed && enabled ? accent.opacity(0.55) : .clear, radius: 14)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 18)
                .stroke(
                    pressed && enabled ? accent : Palette.carbonOutline,
                    lineWidth: pressed && enabled ? 2 : 1
                )
        )
        .scaleEffect(pressed ? 0.92 : 1)
        .animation(.spring(response: 0.18, dampingFraction: 0.55), value: pressed)
        .contentShape(Rectangle())
        .gesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in
                    guard !pressed else { return }   // onChanged repeats while the finger moves
                    pressed = true
                    UIImpactFeedbackGenerator(style: .medium).impactOccurred()
                    onPress()
                }
                .onEnded { _ in
                    pressed = false
                    onRelease()
                }
        )
        .onDisappear {
            // The view going away mid-press must not leave the car running.
            if pressed {
                pressed = false
                onRelease()
            }
        }
    }
}

/// The emergency stop. Always live — even while disconnected, where the view model answers with
/// "Connect to Rocket Pocket first" — because a stop must never be swallowed.
struct EmergencyStopButton: View {
    let onStop: () -> Void
    @State private var pressed = false

    var body: some View {
        VStack(spacing: 3) {
            Image(systemName: "hand.raised.fill")
                .font(.system(size: 24, weight: .bold))
            Text("EMERGENCY\nSTOP")
                .font(.system(size: 9, weight: .semibold))
                .multilineTextAlignment(.center)
        }
        .foregroundColor(Palette.textPrimary)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(
            Circle()
                .fill(pressed ? Palette.neonRed : Palette.neonRedDeep)
                .shadow(color: Palette.neonRed.opacity(pressed ? 0.7 : 0.3), radius: pressed ? 18 : 8)
        )
        .overlay(Circle().stroke(Palette.neonRed, lineWidth: pressed ? 4 : 3))
        .scaleEffect(pressed ? 0.90 : 1)
        .animation(.spring(response: 0.18, dampingFraction: 0.5), value: pressed)
        .contentShape(Circle())
        .gesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in
                    guard !pressed else { return }
                    pressed = true
                    UINotificationFeedbackGenerator().notificationOccurred(.warning)
                    onStop()
                }
                .onEnded { _ in pressed = false }
        )
    }
}

File: RocketPocket/app/src/main/java/com/rocketpocket/car/ui/components/DirectionPad.kt
package com.rocketpocket.car.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.gestures.waitForUpOrCancellation
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material.icons.filled.PanTool
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.rocketpocket.car.bluetooth.Command
import com.rocketpocket.car.ui.theme.NeonCyan
import com.rocketpocket.car.ui.theme.NeonOrange
import com.rocketpocket.car.ui.theme.NeonRed
import com.rocketpocket.car.ui.theme.NeonRedDeep
import com.rocketpocket.car.ui.theme.TextPrimary

/**
 * One cell of the 3x3 pad. The arrow icon is a single upward arrow rotated into place,
 * which keeps all eight directions visually identical.
 */
private data class Direction(
    val command: Char,
    val label: String,
    val rotation: Float,
    val accent: Color,
)

private val Directions = listOf(
    Direction(Command.FORWARD_LEFT, "FWD LEFT", -45f, NeonOrange),
    Direction(Command.FORWARD, "FORWARD", 0f, NeonCyan),
    Direction(Command.FORWARD_RIGHT, "FWD RIGHT", 45f, NeonOrange),
    Direction(Command.LEFT, "LEFT", -90f, NeonCyan),
    Direction(Command.RIGHT, "RIGHT", 90f, NeonCyan),
    Direction(Command.BACKWARD_LEFT, "BACK LEFT", -135f, NeonOrange),
    Direction(Command.BACKWARD, "BACKWARD", 180f, NeonCyan),
    Direction(Command.BACKWARD_RIGHT, "BACK RIGHT", 135f, NeonOrange),
)

private fun direction(command: Char): Direction = Directions.first { it.command == command }

/**
 * The eight-way drive pad with the emergency stop occupying the centre cell.
 *
 * Rows read: forward-left / forward / forward-right, then left / STOP / right,
 * then back-left / backward / back-right.
 */
@Composable
fun DirectionPad(
    enabled: Boolean,
    onDirectionPressed: (Char) -> Unit,
    onDirectionReleased: () -> Unit,
    onEmergencyStop: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // A 3x3 grid of nine equal cells is square by nature. Letting it stretch to fill a wide
    // tablet half turned every button into a flat rectangle, so the grid is pinned to a square
    // sized by the available height and centred in whatever width it is given.
    Box(modifier = modifier, contentAlignment = Alignment.Center) {
        PadGrid(
            enabled = enabled,
            onDirectionPressed = onDirectionPressed,
            onDirectionReleased = onDirectionReleased,
            onEmergencyStop = onEmergencyStop,
            modifier = Modifier
                .fillMaxHeight()
                .aspectRatio(1f, matchHeightConstraintsFirst = true),
        )
    }
}

@Composable
private fun PadGrid(
    enabled: Boolean,
    onDirectionPressed: (Char) -> Unit,
    onDirectionReleased: () -> Unit,
    onEmergencyStop: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        PadRow(
            left = Command.FORWARD_LEFT,
            right = Command.FORWARD_RIGHT,
            centreCommand = Command.FORWARD,
            enabled = enabled,
            onDirectionPressed = onDirectionPressed,
            onDirectionReleased = onDirectionReleased,
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f),
        )

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            DirectionCell(
                direction = direction(Command.LEFT),
                enabled = enabled,
                onDirectionPressed = onDirectionPressed,
                onDirectionReleased = onDirectionReleased,
                modifier = Modifier
                    .weight(1f)
                    .fillMaxSize(),
            )
            EmergencyStopButton(
                onStop = onEmergencyStop,
                modifier = Modifier
                    .weight(1f)
                    .fillMaxSize(),
            )
            DirectionCell(
                direction = direction(Command.RIGHT),
                enabled = enabled,
                onDirectionPressed = onDirectionPressed,
                onDirectionReleased = onDirectionReleased,
                modifier = Modifier
                    .weight(1f)
                    .fillMaxSize(),
            )
        }

        PadRow(
            left = Command.BACKWARD_LEFT,
            right = Command.BACKWARD_RIGHT,
            centreCommand = Command.BACKWARD,
            enabled = enabled,
            onDirectionPressed = onDirectionPressed,
            onDirectionReleased = onDirectionReleased,
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f),
        )
    }
}

@Composable
private fun PadRow(
    left: Char,
    centreCommand: Char,
    right: Char,
    enabled: Boolean,
    onDirectionPressed: (Char) -> Unit,
    onDirectionReleased: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        listOf(left, centreCommand, right).forEach { command ->
            DirectionCell(
                direction = direction(command),
                enabled = enabled,
                onDirectionPressed = onDirectionPressed,
                onDirectionReleased = onDirectionReleased,
                modifier = Modifier
                    .weight(1f)
                    .fillMaxSize(),
            )
        }
    }
}

@Composable
private fun DirectionCell(
    direction: Direction,
    enabled: Boolean,
    onDirectionPressed: (Char) -> Unit,
    onDirectionReleased: () -> Unit,
    modifier: Modifier = Modifier,
) {
    ControlButton(
        label = direction.label,
        icon = Icons.Filled.ArrowUpward,
        accent = direction.accent,
        enabled = enabled,
        iconRotation = direction.rotation,
        onPress = { onDirectionPressed(direction.command) },
        onRelease = onDirectionReleased,
        modifier = modifier,
    )
}

/**
 * The emergency stop. It is always live — even while disconnected, where the view model
 * answers with "Connect to Rocket Pocket first" — because a stop must never be swallowed.
 */
@Composable
fun EmergencyStopButton(
    onStop: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val haptics = LocalHapticFeedback.current
    val currentOnStop by rememberUpdatedState(onStop)
    var pressed by remember { mutableStateOf(false) }

    val scale by animateFloatAsState(
        targetValue = if (pressed) 0.90f else 1f,
        animationSpec = spring(dampingRatio = 0.5f, stiffness = 900f),
        label = "emergencyStopScale",
    )

    Box(
        modifier = modifier
            .scale(scale)
            .background(
                brush = Brush.radialGradient(
                    listOf(if (pressed) NeonRed else NeonRedDeep, NeonRedDeep.copy(alpha = 0.85f)),
                ),
                shape = CircleShape,
            )
            .border(
                width = if (pressed) 4.dp else 3.dp,
                color = NeonRed,
                shape = CircleShape,
            )
            .pointerInput(Unit) {
                awaitEachGesture {
                    val down = awaitFirstDown(requireUnconsumed = false)
                    down.consume()
                    pressed = true
                    haptics.performHapticFeedback(HapticFeedbackType.LongPress)
                    currentOnStop()

                    val up = waitForUpOrCancellation()
                    up?.consume()
                    pressed = false
                }
            },
        contentAlignment = Alignment.Center,
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.Center,
        ) {
            Icon(
                imageVector = Icons.Filled.PanTool,
                contentDescription = "Emergency Stop",
                tint = TextPrimary,
                modifier = Modifier.size(28.dp),
            )
            Text(
                text = "EMERGENCY\nSTOP",
                style = MaterialTheme.typography.labelSmall,
                color = TextPrimary,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(top = 3.dp),
            )
        }
    }
}
